`Resolver` type from webpack/enhanced-resolve are not compatible

# Bug report

`import("webpack").Resolver` should be compatible with `import("enhanced-resolve").Resolver`

[webpack/types.d.ts](https://github.com/webpack/webpack/blob/6be4065ade1e252c1d8dcba4af0f43e32af1bdc1/types.d.ts#L4528)

```ts
arg1?: (string | Buffer)[] | (typeof Dirent)[]
```

[enhanced-resolve/types.d.ts](https://github.com/webpack/enhanced-resolve/blob/a998c7d218b7a9ec2461fc4fddd1ad5dd7687485/types.d.ts#L60)

```ts
arg1?: (string | Buffer)[] | Dirent[]
```

**What is the current behavior?**

```
Error: src/createWebpackConfig.ts(47,17): error TS2322: Type 'ProjectReferencesPlugin' is not assignable to type 'false | "" | 0 | ResolvePluginInstance | "..." | null | undefined'.
  Type 'ProjectReferencesPlugin' is not assignable to type 'ResolvePluginInstance'.
    Types of property 'apply' are incompatible.
      Type '(resolver: import("/home/runner/work/buncho/buncho/node_modules/enhanced-resolve/types").Resolver) => void' is not assignable to type '(resolver: import("/home/runner/work/buncho/buncho/node_modules/webpack/types").Resolver) => void'.
        Types of parameters 'resolver' and 'resolver' are incompatible.
          Type 'import("/home/runner/work/buncho/buncho/node_modules/webpack/types").Resolver' is not assignable to type 'import("/home/runner/work/buncho/buncho/node_modules/enhanced-resolve/types").Resolver'.
            The types of 'fileSystem.readdir' are incompatible between these types.
              Type '(arg0: string, arg1?: "ascii" | "utf8" | "utf-8" | "utf1[6](https://github.com/neetly/buncho/actions/runs/5437834022/jobs/9888646844?pr=1358#step:9:7)le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | ((arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | ... 1 more ... | undefined) => void) | ReaddirOptions | "buffer" | null | undefined, arg2?: ((arg0?: ErrnoE...' is not assignable to type '(arg0: string, arg1?: "ascii" | "utf[8](https://github.com/neetly/buncho/actions/runs/5437834022/jobs/9888646844?pr=1358#step:9:9)" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | "buffer" | ((arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | ... 1 more ... | undefined) => void) | ReaddirOptions | null | undefined, arg2?: ((arg0?: ErrnoE...'.
                Types of parameters 'arg1' and 'arg1' are incompatible.
                  Type '"ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | "buffer" | ((arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | ... 1 more ... | undefined) => void) | ReaddirOptions | null | undefined' is not assignable to type '"ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | ((arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | ... 1 more ... | undefined) => void) | ReaddirOptions | "buffer" | null | undefined'.
                    Type '(arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | Dirent[] | undefined) => void' is not assignable to type '"ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | ((arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | ... 1 more ... | undefined) => void) | ReaddirOptions | "buffer" | null | undefined'.
                      Type '(arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | Dirent[] | undefined) => void' is not assignable to type '(arg0?: ErrnoException | null | undefined, arg1?: (string | Buffer)[] | (typeof Dirent)[] | undefined) => void'.
                        Types of parameters 'arg1' and 'arg1' are incompatible.
                          Type '(string | Buffer)[] | (typeof Dirent)[] | undefined' is not assignable to type '(string | Buffer)[] | Dirent[] | undefined'.
                            Type '(typeof Dirent)[]' is not assignable to type '(string | Buffer)[] | Dirent[] | undefined'.
                              Type '(typeof Dirent)[]' is not assignable to type '(string | Buffer)[]'.
                                Type 'typeof Dirent' is not assignable to type 'string | Buffer'.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

There should be no type error.

**Other relevant information:**
webpack version: 5.88.1

--- a/package.json
+++ b/package.json
@@ -98,7 +98,7 @@
     "style-loader": "^2.0.0",
     "terser": "^5.26.0",
     "toml": "^3.0.0",
-    "tooling": "webpack/tooling#v1.23.0",
+    "tooling": "webpack/tooling#v1.23.1",
     "ts-loader": "^9.4.2",
     "typescript": "^5.0.4",
     "url-loader": "^4.1.0",

--- a/types.d.ts
+++ b/types.d.ts
@@ -4571,7 +4571,7 @@ declare interface FileSystem {
 			| "binary"
 			| ((
 					arg0?: null | NodeJS.ErrnoException,
-					arg1?: (string | Buffer)[] | (typeof Dirent)[]
+					arg1?: (string | Buffer)[] | Dirent[]
 			  ) => void)
 			| ReaddirOptions
 			| "utf-8"
@@ -4582,7 +4582,7 @@ declare interface FileSystem {
 			| "buffer",
 		arg2?: (
 			arg0?: null | NodeJS.ErrnoException,
-			arg1?: (string | Buffer)[] | (typeof Dirent)[]
+			arg1?: (string | Buffer)[] | Dirent[]
 		) => void
 	) => void;
 	readJson?: {
@@ -6896,7 +6896,7 @@ declare interface LazyCompilationDefaultBackendOptions {
 	/**
 	 * Specifies where to listen to from the server.
 	 */
-	listen?: number | ListenOptions | ((server: typeof Server) => void);
+	listen?: number | ListenOptions | ((server: Server) => void);
 
 	/**
 	 * Specifies the protocol the client should use to connect to the server.
@@ -6909,7 +6909,7 @@ declare interface LazyCompilationDefaultBackendOptions {
 	server?:
 		| ServerOptionsImport<typeof IncomingMessage>
 		| ServerOptionsHttps<typeof IncomingMessage, typeof ServerResponse>
-		| (() => typeof Server);
+		| (() => Server);
 }
 
 /**

